src/sync/barrier.rs

```diff
@@ -64,22 +66,35 @@ impl Barrier {
             state.leader = Some(me);
         }
 
+        // Update the barrier's clock with the clock of this thread
+        ExecutionState::with(|s| {
+            let clock = s.increment_clock();
+            state.clock.update(clock);
+        });
+
         if state.waiters.len() + 1 < state.bound {
             // Block current thread
             assert!(state.waiters.insert(me)); // current thread shouldn't already be in the set
             ExecutionState::with(|s| s.current_mut().block());
             trace!(leader=?state.leader, waiters=?state.waiters, "blocked on barrier {:?}", self);
         } else {
             trace!(leader=?state.leader, waiters=?state.waiters, "releasing waiters on barrier {:?}", self);
-            // Unblock all waiters
-            for tid in state.waiters.drain() {
-                debug_assert_ne!(tid, me);
-                ExecutionState::with(|s| {
+            // Unblock each waiter, updating its clock with the barrier's clock
+            let clock = state.clock.clone();
+            ExecutionState::with(|s| {
+                for tid in state.waiters.drain() {
+                    debug_assert_ne!(tid, me);
                     let t = s.get_mut(tid);
                     debug_assert!(t.blocked());
+                    t.clock.increment(tid);
+                    t.clock.update(&clock);
                     t.unblock();
```

Comment:
I wonder if there's some way to make this API less error-prone by making `unblock` take a clock as an argument and do the required clock updates itself.

Reply:
Problem is that there are places where we call unblock only in some situations.  Example: when a sender sends a message on an empty channel, if there's a waiting receiver, it unblocks the receiver.  If we passed in the sender clock to unblock, we'd update the receiver clock with it, which is fine.  But then the receiver code that removes the message will also update the clock from the message, which is redundant now. And we can't avoid updating the receiver with the clock, because that's needed in the case where the sender didn't wake the receiver.

Also in some places (e.g., Condvar::notify_one), we unblock a bunch of threads, but the first one blocks the others.  In this case, we don't want to update the clocks of anyone except the winner.

Reply:
That makes sense. I think in practice these are likely to be pretty stable anyway, and new primitives are unlikely to fit neatly into some predefined clock pattern.